### PR TITLE
EOS-15345: Script for enabling code coverage for s3server

### PR DESCRIPTION
**Pull request by Shreya-18**

### CODE REVIEW

File: scripts/coverage/run_all_coverage.sh [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        *)
+            usage
+            ;;
+    esac
+done
+shift $((OPTIND-1))
+
+mkdir -p $DES_DIR
+run_coverage
+upload_report
```
Comment (nileshgovande) [inline scripts/coverage/run_all_coverage.sh:92]: Newline at EOF

File: scripts/coverage/run_all_coverage.sh [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For any questions about this software or licensing,
+# please email opensource@seagate.com or cortx-questions@seagate.com.
+#
+
+set -e
+set +o posix
```
Comment (nileshgovande) [inline scripts/coverage/run_all_coverage.sh:22]: Add deletion step

Comment (nileshgovande) [reply scripts/coverage/run_all_coverage.sh:22]: Better add it at start of individual script

File: scripts/coverage/python_coverage.sh [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    # Generate xml report
+    coverage xml -o "$DES_DIR/s3server_python_coverage.xml"
+    popd
+}
+
+check_report() {
+    test -f "$DES_DIR/s3server_python_coverage.xml" || die "Failed to generate the Coverage report."
+}
+
+run_ut
```
Comment (nileshgovande) [inline scripts/coverage/python_coverage.sh:68]: make an explicit call to check perquisites()
Also write a comment as this is 'main' start of script

Comment (nileshgovande) [reply scripts/coverage/python_coverage.sh:68]: all 4 scripts

File: scripts/coverage/run_all_coverage.sh [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+# variables
+SCRIPT_PATH=$(readlink -f "$0")
+SCRIPT_DIR=$(dirname "$SCRIPT_PATH")
+SRC_DIR="$(dirname "$(dirname "$SCRIPT_DIR" )")"
+#default codacy project token
+CODACY_PROJECT_TOKEN=7a64ef82e1a049ab898fe838f1320594
+BRANCH="main"
+
+export DES_DIR="/root/code_coverage"
```
Comment (nileshgovande) [inline scripts/coverage/run_all_coverage.sh:32]: The default should be within the project itself - let's not do it in the root

Comment (nileshgovande) [reply scripts/coverage/run_all_coverage.sh:32]: Try - /root/workspace/cortx-s3server/.coverage 
Also - take this as a parameter if someone wants overwrite

File: scripts/coverage/run_all_coverage.sh [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+upload_report() {
+    export CODACY_PROJECT_TOKEN
+    export CODACY_API_BASE_URL=https://api.codacy.com
+
+    git checkout $BRANCH
+
+    codacy_coverage_url="https://coverage.codacy.com/get.sh"
+
+    # Upload cpp coverage report
```
Comment (nileshgovande) [inline scripts/coverage/run_all_coverage.sh:58]: the same report pushed - pls write reason

File: scripts/coverage/cpp_coverage.sh [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For any questions about this software or licensing,
+# please email opensource@seagate.com or cortx-questions@seagate.com.
+#
+
+# This script helps to generate coverage report (.xml) for cpp code.
```
Comment (nileshgovande) [inline scripts/coverage/cpp_coverage.sh:21]: Pls write comments on individual functions

File: scripts/coverage/cpp_coverage.sh [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    find . -name murmur* -exec rm -f {} \;
+    find . -name jsoncpp* -exec rm -f {} \;
+    find . -name base64* -exec rm -f {} \;
+
+    # Generate coverage report
+    lcov -t "s3server coverage result" --no-external -b $SRC_DIR -o s3server_coverage.info -c -d .
+    popd
+}
+
+run_coverage() {
```
Comment (nileshgovande) [inline scripts/coverage/cpp_coverage.sh:97]: Make explicit

File: scripts/coverage/cpp_coverage.sh [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+LCOV_DIR="/root/lcovlog"
+
+die() {
+  echo "${*}"
+  exit 1
+}
+
+check_prerequisite() {
+    # install lcov rpm if not present
+    rpm -q lcov &> /dev/null || yum install lcov
```
Comment (nileshgovande) [inline scripts/coverage/cpp_coverage.sh:40]: add -y parameter to yum

File: scripts/coverage/cpp_coverage.sh [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+set -e
+
+#variables
+SCRIPT_PATH=$(readlink -f "$0")
+SCRIPT_DIR=$(dirname "$SCRIPT_PATH")
+SRC_DIR="$(dirname "$(dirname "$SCRIPT_DIR" )")"
+DES_DIR=${DES_DIR:-"/root/code_coverage"}
+BUILD_FILE="./BUILD.template"
+LCOV_DIR="/root/lcovlog"
```
Comment (nileshgovande) [inline scripts/coverage/cpp_coverage.sh:31]: Move to project

File: scripts/coverage/cpp_coverage.sh [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+run_ut() {
+    echo "Executing UTs"
+    pushd $SRC_DIR
+    # Update the BUILD.template file   
+    add_coverage_options
+    # Run UTs
+    set +e
+    ./jenkins-build.sh
```
Comment (nileshgovande) [inline scripts/coverage/cpp_coverage.sh:64]: try reducing time

File: scripts/coverage/run_all_coverage.sh [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+#
+
+set -e
+set +o posix
+
+# variables
+SCRIPT_PATH=$(readlink -f "$0")
+SCRIPT_DIR=$(dirname "$SCRIPT_PATH")
+SRC_DIR="$(dirname "$(dirname "$SCRIPT_DIR" )")"
+#default codacy project token
```
Comment (nileshgovande) [inline scripts/coverage/run_all_coverage.sh:28]: s3 token

File: scripts/coverage/python_coverage.sh (hunk 3 above)
Comment (Shreya-18) [reply scripts/coverage/python_coverage.sh:84]: Done

File: scripts/coverage/cpp_coverage.sh (hunk 7 above)
Comment (Shreya-18) [reply scripts/coverage/cpp_coverage.sh:87]: done

File: scripts/coverage/cpp_coverage.sh (hunk 6 above)
Comment (Shreya-18) [reply scripts/coverage/cpp_coverage.sh:21]: done

File: scripts/coverage/run_all_coverage.sh (hunk 2 above)
Comment (Shreya-18) [reply scripts/coverage/run_all_coverage.sh:22]: Sure, added in all child scripts.

File: scripts/coverage/run_all_coverage.sh (hunk 5 above)
Comment (Shreya-18) [reply scripts/coverage/run_all_coverage.sh:64]: done

File: scripts/coverage/run_all_coverage.sh (hunk 1 above)
Comment (Shreya-18) [reply scripts/coverage/run_all_coverage.sh]: done

File: scripts/coverage/run_all_coverage.sh (hunk 4 above)
Comment (Shreya-18) [reply scripts/coverage/run_all_coverage.sh]: added and changed to project directory.
Done

File: scripts/coverage/cpp_coverage.sh (hunk 8 above)
Comment (Shreya-18) [reply scripts/coverage/cpp_coverage.sh]: done

File: scripts/coverage/cpp_coverage.sh (hunk 9 above)
Comment (Shreya-18) [reply scripts/coverage/cpp_coverage.sh]: done

File: scripts/coverage/run_all_coverage.sh (hunk 11 above)
Comment (Shreya-18) [reply scripts/coverage/run_all_coverage.sh]: added to comment

**Review by sachitanands — APPROVED** [review]
looks good

**Comment by sachitanands** [comment]
Pre-merge job : http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server-7.8.2003/339/

**Review by nileshgovande — APPROVED** [review]

**Comment by nileshgovande** [comment]
Not part of Jenkins job - Merging

**Pull request merged**